### DISCUSSION TITLE
The contradiction between the q grid of QE ph.x and scf.in of SSCHA.

### DISCUSSION THREAD

**Discussion by LHDphy** [SSCHAcode/python-sscha#129] [Q&A] [answered]
Dear all,

First, please note that the q-grid of QE ph.x must be equal to the expanded cell number of [scf.in](http://scf.in/) in SSCHA. For example, if a 4x4x1 q-grid is used in ph.x, there will be a 4x4x1 supercell in the [scf.in](http://scf.in/) of SSCHA.

However, a problem arises when we use QE to calculate phonon dispersion - the convergence of the q-grid must be tested. The minimum 12x12x1 q-grid must be used, otherwise, the phonon dispersions may differ from those reported in other papers.

**If a unit cell has five atoms, it is hard to imagine how massive the computational workload will be!**

Is it possible to solve this problem by modifying the source code?

Thank you for your response in advance.

Sincerely,

**Comment by mesonepigreco** [comment, marked as answer]
Dear LHDphy,
The q-grid employed in the phonon calculation corresponds to the dimension of the supercell of the SSCHA calculation. And it is also true that the supercell size must be tested for convergence.
However, the supercell size necessary to converge an SSCHA calculation is usually much smaller than the one necessary to get an excellent harmonic band structure with phonons so that smaller cells can be employed.

In the end, the optimal size of the supercell depends on the system to be studied and the properties you need to calculate. 

From your grid, I assume you are studying a 2D material for which a 4x4x1 could be too small. However, when computing the anharmonic phonon dispersion, you may still employ a 12x12x1 harmonic q-mesh and combine it with a 4x4x1 sscha mesh.

You can use the interpolate function of CellConstructor as:

```
dyn_harmonic_12x12 = CC.Phonons.Phonons("harm_dyn_12_", nqirr_12)
dyn_harmonic_4x4 = CC.Phonons.Phonons("harm_dyn_4_", nqirr_4)

dyn_sscha_4x4  = CC.Phonons.Phonons("sscha_dyn_4_", nqirr_4)

anharmonic_dyn_12x12  = dyn_sscha_4x4.Interpolate((4,4,1), (12,12,1),  support_dyn_coarse=dyn_harmonic_4x4,  support_dy_fine=dyn_harmonic_12x12)
anharmonic_dyn_12x12.save_qe("sscha_12x12_")
```

In this way, you only interpolate the difference between the harmonic and anharmonic dynamical matrix, which usually requires much lower supercells to converge.

I hope this answer your questions,
All the best,
Lorenzo

**Reply by LHDphy** [reply]
Dear Lorenzo,
Thank you for your response! 

I will try the method you suggested to see if it can solve my problem. 

Once again, thank you very much for your reply!

Sincerely,
LHDphy

**Reply by LHDphy** [reply]
Dear Lorenzo,

Thanks for your reply! I can get the meaning what you expressed. By this ways, we can avoid tedious calculations. Meanwhile, I also try my best to apply your advice.

**However, I can not operate it. I don't know which file to put this code in. **

Can you describe it in more detail? I will greatly appreciate it.

Best Wishes.
LHDphy.

**Reply by mesonepigreco** [reply]
Hi, 

This needs to be done after the full sscha calculation on the 4x4 supercell.

You need:
- Harmonic dynamical matrix in the 4x4
- Harmonic dynamical matrix in the 12x12
- SSCHA auxiliary matrix in the 4x4

Then create a new python script with the following content:

```
import cellconstructor as CC, cellconstructor.Phonons

nqirr_12 = XXX # Here the number of irreducible q points for the 12x12
nqirr_4 = YYY # Here the number of irreducible q points for the 4x4 

dyn_harmonic_12x12 = CC.Phonons.Phonons("harm_dyn_12_", nqirr_12)
dyn_harmonic_4x4 = CC.Phonons.Phonons("harm_dyn_4_", nqirr_4)

dyn_sscha_4x4  = CC.Phonons.Phonons("sscha_dyn_4_", nqirr_4)

anharmonic_dyn_12x12  = dyn_sscha_4x4.Interpolate((4,4,1), (12,12,1),  support_dyn_coarse=dyn_harmonic_4x4,  support_dy_fine=dyn_harmonic_12x12)
anharmonic_dyn_12x12.save_qe("sscha_12x12_")
```

Where you need to substitute the name of the dynamical matrices with those obtained from the aforesaid calculations. Run it with

```
python name_of_the_script.py
```

Hope this helps,
All the best,
Lorenzo

**Reply by LHDphy** [reply]
Thank you!  I will try my best again.
